### DISCUSSION TITLE
storage/mmap: to call mmap.Flush() ?

### DISCUSSION THREAD

**Discussion by AskAlexSharov** [anacrolix/torrent#751] [Ideas]
Hi.

I wonder what will happen in next case: 
- in db stored that piece N is completed
- power off (OS didn't write PageCache to disk)
- on next start App will think that piece N is valid on disk

I don't vote for full ACID here (even that sqlite and boltdb do support transactions). Just wondering: if we can reduce probability of such event. 

Does it make sense to call mmap.Flush at some point? When file fully downloaded, or before file close, or periodically or at somehow else.

**Comment by anacrolix** [comment]
@AskAlexSharov I have considered this in the past. I don't believe mmap flush semantics are standardized, most OS will flush at their whim which is probably best for performance. It's likely that if you have a persistent piece completion database in use, that if it syncs that a piece is complete, but the data isn't actually flushed to disk you could incorrectly treat the piece as complete after dataloss. Perhaps flushing a piece while marking it complete is worthwhile (probably synchronously in that part of the code?). Are you interested in contributing that?

**Reply by AskAlexSharov** [reply]
- fsync (or msync) can be sync and async. 
- can sync after piece write or after file write. 
We will think about contribution.

- What is right way to deal with invalid file? Just client.RemoveTorrent() + client.AddTorrent() (and it will validate things)? (right now I do delete anacrolix's db if some hash doesn't match).

**Reply by anacrolix** [reply]
> * fsync (or msync) can be sync and async.
> * can sync after piece write or after file write.
>   We will think about contribution.

👍🏻 

> * What is right way to deal with invalid file? Just client.RemoveTorrent() + client.AddTorrent() (and it will validate things)? (right now I do delete anacrolix's db if some hash doesn't match).

Call Piece or Torrent VerifyData or the client won't know the data is invalid (unless it gets an error while trying to read). A File equivalent could be added, but it would just be doing the calculations on which pieces and calling Piece.VerifyData for those.
